0049-protocol-versioning.md

```diff
@@ -0,0 +1,42 @@
+
+Compatibility of the networking protocol can be thought of as the set of all RPC messages, wires types, gossip topics, p2p feature set, and participation rules. Participation rules include any rule in the protocol that regards certain behavior as malicious or otherwise not permitted. This ranges from consensus details to bannable offenses and even verification logic for gossip messages. It's important the capture all of this under the umbrella of networking protocol compatibility since divergences in these details can lead to unintended forks on the chain.
+
+To label versions of our daemon software, we will use the following versioning convention, inspired by semver: `<txn_version>.<net_version>.<impl_version>`. In this setup, the version of the transaction system is the most dominate version, since any updates to it necessitate a hardfork, and when reasoning about the logic of the chain state, it is the most significant version number. The prefix of the `<txn_version>.<net_version>` uniquely identifies a particular hard fork version, since the pair of those 2 versions determines the full compatability of an implementation. The leftover `<impl_version>` is retained for the usage of individual daemon implementations to use however they see fit. We leave this detail to each implementor of the daemon, as the meaning of these versions is intended to be specific to the implementation.
+
+For the existing daemon implementation in OCaml, which is maintained in the `MinaProtocol/mina` repository, we will use the following versioning convention for the `<impl_version>`: `<api_version>-<patch_version>`. Here, the `<api_version>` will be used to denotate any breaking API changes for user-facing APIs the daemon supports (CLI, GraphQL, Archive Format). Whenever we add, remove, deprecate, or modify the interface of existing APIs, this version must be incremented. The `<patch_version>` is used in the same way the patch version is utilized in semver: to signify that there are new backwards compatible bug fixes or improvements. We may also add an additional suffix to the `<impl_version>` if there is a different variant of the artifact for that version. For instance, if we were testing some optimizations behind a feature flag, but weren't ready to ship it to the stable version of the software (or didn't want to for some reason), then we could maintain a variant build for that artifact by appending a `-opt` prefix to the version.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback of this plan is that it requires additional review when tagging the version for a new release of the software. However, this appears to be a necessary and an important step in the process, and hopefully will lead us to a world where we are developing the protocol separately from the implementation, rather than assigning a protocol version to an implementation retroactively upon release.
```

Comment:
This would be really nice to have, but I the only way I can think to do this while maintaining a reasonable "public" versioning sequence (such that berkeley is v2) would require maintaining 2 parallel version sequences (one for internal and one for external). That would probably be difficult to communicate around and requires more thought. Let's revisit that idea outside of this RFC if we want to do something like this.